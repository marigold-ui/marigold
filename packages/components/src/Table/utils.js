/**
 * Map `react-aria` props to ours (no "is"-prefix)
 */
export const mapCheckboxProps = ({
  checkboxProps: {
    isIndeterminate,
    isSelected,
    isDisabled,
    defaultSelected,
    ...rest
  },
}) => {
  const checkboxProps = {
    disabled: isDisabled,
    checked: isSelected,
    defaultChecked: defaultSelected,
    indeterminate: isIndeterminate,
    ...rest,
  };
  return { checkboxProps };
};
//# sourceMappingURL=utils.js.map
