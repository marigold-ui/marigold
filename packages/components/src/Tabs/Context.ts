import { ComponentStyleParts } from '@marigold/system';
import { createContext, useContext } from 'react';

export interface TabContextProps {
  styles: ComponentStyleParts<['tabs', 'tab', 'tabPanel']>;
}

export const TabContext = createContext<TabContextProps>({} as any);
export const useTabContext = () => useContext(TabContext);
