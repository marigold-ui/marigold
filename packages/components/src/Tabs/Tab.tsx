import React from 'react';
import { useRef } from 'react';

import { useFocus, useHover } from '@react-aria/interactions';
import { useTab } from '@react-aria/tabs';
import { mergeProps } from '@react-aria/utils';

import { TabListState } from '@react-stately/tabs';

import { Node } from '@react-types/shared';

import { cn, useStateProps } from '@marigold/system';

import { useTabContext } from './Context';

export interface TabProps {
  item: Node<object>;
  state: TabListState<object>;
}

export const Tab = ({ item, state }: TabProps) => {
  const { key, rendered } = item;
  const ref = useRef(null);
  const { tabProps, isSelected } = useTab({ key }, state, ref);
  const disabled = tabProps['aria-disabled'];
  const { hoverProps, isHovered } = useHover({
    isDisabled: (disabled as boolean) || isSelected,
  });
  const { focusProps } = useFocus({});
  const stateProps = useStateProps({ active: isSelected, hover: isHovered });
  const { classNames } = useTabContext();

  return (
    <div
      className={cn(
        'flex cursor-pointer justify-center aria-disabled:cursor-not-allowed',
        classNames.tab
      )}
      {...mergeProps(tabProps, stateProps, focusProps, hoverProps)}
      ref={ref}
    >
      {rendered}
    </div>
  );
};
