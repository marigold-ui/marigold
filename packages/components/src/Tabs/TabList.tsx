import { TabList } from 'react-aria-components';
import type RAC from 'react-aria-components';

import { GapSpaceProp, cn, gapSpace } from '@marigold/system';

import { useTabContext } from './Context';

// props
// ----------------------
export interface TabListProps
  extends Omit<RAC.TabListProps<Object>, 'className' | 'style'>,
    GapSpaceProp {}

// component
// ----------------------
const _TabList = ({ space = 2, ...props }: TabListProps) => {
  const { classNames } = useTabContext();
  return (
    <TabList
      {...props}
      className={cn('flex', gapSpace[space], classNames.tabsList)}
    >
      {props.children}
    </TabList>
  );
};

export { _TabList as TabList };
