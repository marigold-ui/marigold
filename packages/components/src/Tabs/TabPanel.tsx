import React from 'react';
import { useRef } from 'react';
import { AriaTabPanelProps, useTabPanel } from '@react-aria/tabs';
import { TabListState } from '@react-stately/tabs';
import { useTabContext } from './Context';
import { cn } from '@marigold/system';

export interface TabPanelProps extends AriaTabPanelProps {
  state: TabListState<object>;
}

export const TabPanel = ({ state, ...props }: TabPanelProps) => {
  const ref = useRef(null);
  const { tabPanelProps } = useTabPanel(props, state, ref);
  const selectedItemProps = state.selectedItem?.props;
  const { classNames } = useTabContext();

  return (
    <div className={cn(classNames.tabpanel)} ref={ref} {...tabPanelProps}>
      {selectedItemProps?.children}
    </div>
  );
};
