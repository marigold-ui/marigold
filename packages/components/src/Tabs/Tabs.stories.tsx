import type { Meta, StoryObj } from '@storybook/react';

import { Tabs, TabsProps } from './Tabs';

const meta = {
  title: 'Components/Tabs',
  argTypes: {
    disabled: {
      control: { type: 'boolean' },
      description: 'Disable all the tabs',
    },
    selectedKey: {
      control: { type: 'text' },
      description: 'Select the active tab',
    },
    size: {
      control: { type: 'select' },
      description: 'Select the size of the tab controller',
      options: ['small', 'medium', 'large'],
    },
    keyboardActivation: {
      defaultValue: 'automatic',
      control: { type: 'radio' },
      options: ['automatic', 'manual'],
      description:
        'Select whether tabs are activated automatically on focus or manually.',
    },
  },
} satisfies Meta<TabsProps>;

export default meta;

export const Basic: StoryObj<typeof Tabs> = {
  render: args => {
    return (
      <Tabs aria-label="tabs" {...args}>
        <Tabs.List aria-label="Input settings">
          <Tabs.Item id="mouse">Mouse Settings</Tabs.Item>
          <Tabs.Item id="keyboard">Keyboard Settings</Tabs.Item>
          <Tabs.Item id="gamepad">Gamepad Settings</Tabs.Item>
        </Tabs.List>
        <Tabs.TabPanel id="mouse">
          Adjust the sensitivity, acceleration, and button assignments for your
          mouse.
        </Tabs.TabPanel>
        <Tabs.TabPanel id="keyboard">
          Customize the key bindings and input behavior for your keyboard.
        </Tabs.TabPanel>
        <Tabs.TabPanel id="gamepad">
          Configure the controls, dead zones, and vibration settings for your
          gamepad.
        </Tabs.TabPanel>
      </Tabs>
    );
  },
};

export const WithDisabledKeys: StoryObj<typeof Tabs> = {
  render: args => {
    return (
      <Tabs aria-label="tabs" disabledKeys={['private']} {...args}>
        <Tabs.List aria-label="User Preferences">
          <Tabs.Item id="profile">profile</Tabs.Item>
          <Tabs.Item id="notifications">notifications</Tabs.Item>
          <Tabs.Item id="private">private</Tabs.Item>
        </Tabs.List>
        <Tabs.TabPanel id="profile">
          This panel displays your profile settings. You can upload a profile
          picture, write a brief bio to introduce yourself, and update other
          personal details. Show the world who you are and make a memorable
          impression on other platform users.
        </Tabs.TabPanel>
        <Tabs.TabPanel id="notifications">
          Here, you'll find settings related to notifications. Choose how you
          want to be notified about new messages, friend requests, and other
          important updates. Stay connected and up-to-date with the latest
          activities happening on the platform.
        </Tabs.TabPanel>
        <Tabs.TabPanel id="private">
          The Privacy panel provides options to safeguard your personal
          information and control your privacy. Decide who can view your
          profile, set visibility preferences for your posts and photos, and
          manage your data. Enjoy peace of mind knowing that you have full
          control over your privacy on the platform.
        </Tabs.TabPanel>
      </Tabs>
    );
  },
};

export const WithSelectedTab: StoryObj<typeof Tabs> = {
  render: args => {
    return (
      <Tabs aria-label="tabs" selectedKey={'settings'} {...args}>
        <Tabs.List aria-label="Navigation">
          <Tabs.Item id="home">Home</Tabs.Item>
          <Tabs.Item id="settings">Settings</Tabs.Item>
          <Tabs.Item id="messages">Messages</Tabs.Item>
        </Tabs.List>
        <Tabs.TabPanel id="home">
          Welcome to the Home tab! This is where you can find your personalized
          feed, updates from your friends, and recent activity on the platform.
          Stay connected and explore the latest content tailored to your
          interests.
        </Tabs.TabPanel>
        <Tabs.TabPanel id="settings">
          You're currently in the Settings tab. Here, you can customize your
          account preferences, manage privacy settings, and update your
          notification preferences. Personalize your experience and make the
          platform work exactly how you want it to.
        </Tabs.TabPanel>
        <Tabs.TabPanel id="messages">
          Check out your Messages tab to stay in touch with your contacts. Send
          and receive messages, engage in conversations, and share interesting
          content with your network. Connect with friends, colleagues, and
          like-minded individuals in a private and secure environment.
        </Tabs.TabPanel>
      </Tabs>
    );
  },
};
