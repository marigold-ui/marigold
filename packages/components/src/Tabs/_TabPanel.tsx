import { useRef } from 'react';

import { AriaTabPanelProps, useTabPanel } from '@react-aria/tabs';

import { TabListState } from '@react-stately/tabs';

import { cn } from '@marigold/system';

import { useTabContext } from './Context';

export interface TabPanelProps extends AriaTabPanelProps {
  state: TabListState<object>;
}

export const TabPanel = ({ state, ...props }: TabPanelProps) => {
  const ref = useRef(null);
  const { tabPanelProps } = useTabPanel(props, state, ref);
  const selectedItemProps = state.selectedItem?.props;
  const { classNames } = useTabContext();

  return (
    <div className={cn(classNames.tabpanel)} ref={ref} {...tabPanelProps}>
      {selectedItemProps?.children}
    </div>
  );
};
