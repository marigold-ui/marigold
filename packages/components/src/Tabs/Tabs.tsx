import React from 'react';
import { useTabList } from '@react-aria/tabs';
import { GapSpaceProp, cn, useClassNames, gapSpace } from '@marigold/system';
import { useTabListState } from '@react-stately/tabs';
import { useRef } from 'react';
import { AriaTabListProps } from '@react-types/tabs';

import { Tab } from './Tab';
import { TabPanel } from './TabPanel';
import { Item } from '@react-stately/collections';
import { TabContext } from './Context';

interface TabsProps
  extends Omit<AriaTabListProps<object>, 'orientation' | 'isDisabled'>,
    GapSpaceProp {
  size?: 'small' | 'medium' | 'large';
  disabled?: boolean;
  variant?: string;
}
export const Tabs = ({
  space = 2,
  size = 'medium',
  disabled,
  variant,
  ...rest
}: TabsProps) => {
  const ref = useRef(null);
  const props: AriaTabListProps<object> = {
    isDisabled: disabled,
    ...rest,
  };
  const state = useTabListState(props);
  const { tabListProps } = useTabList(props, state, ref);

  // We don't have variant for now , it is only for future use
  const classNames = useClassNames({
    component: 'Tabs',
    size,
    variant,
  });

  return (
    <TabContext.Provider value={{ classNames }}>
      <div
        className={cn('flex', gapSpace[space], classNames.tabs)}
        {...tabListProps}
        ref={ref}
      >
        {[...state.collection].map(item => {
          return <Tab key={item.key} item={item} state={state} />;
        })}
      </div>
      <TabPanel key={state.selectedItem?.key} state={state} />
    </TabContext.Provider>
  );
};

Tabs.Item = Item;
