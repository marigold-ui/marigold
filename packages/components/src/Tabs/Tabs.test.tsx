import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import { Theme } from '@marigold/system';
import { Tabs } from './Tabs';
import { cva } from 'class-variance-authority';
import { setup } from '../test.utils';

const theme: Theme = {
  name: 'tabs test',
  components: {
    Tabs: {
      tabs: cva('mb-[10px]'),
      tab: cva(
        [
          'min-h-[40px]',
          'data-[hover]:text-tabs-tab-text data-[hover]:border-b-tabs-tab-hover data-[hover]:border-b-8 data-[hover]:border-solid',
          ' disabled:text-tabs-tab-disabled',
          ' selected:border-b-primary-600  selected:border-b-8  selected:border-solid ',
        ],
        {
          variants: {
            size: {
              small: 'px-1 pb-1',
              medium: 'px-2 pb-2 text-lg',
              large: 'px-4 pb-4 text-2xl',
            },
          },
        }
      ),
    },
  },
};

const { render } = setup({ theme });

test('rendering content correctly', () => {
  render(
    <Tabs>
      <Tabs.Item title="tab1" key={1}>
        tab-1 content
      </Tabs.Item>
      <Tabs.Item title="tab2" key={2}>
        tab-2 content
      </Tabs.Item>
      <Tabs.Item title="tab3" key={3}>
        tab-3 content
      </Tabs.Item>
    </Tabs>
  );

  // rendering the tab controller
  expect(screen.getByText('tab1')).toBeInTheDocument();
  // rendering tabpanel
  expect(screen.getByText(/tab-1/)).toBeInTheDocument();
});

test('Supporting default size', () => {
  render(
    <Tabs>
      <Tabs.Item key={1} title="tab">
        tab content
      </Tabs.Item>
    </Tabs>
  );
  expect(screen.getByText('tab').className).toMatchInlineSnapshot(
    `"flex cursor-pointer justify-center aria-disabled:cursor-not-allowed min-h-[40px] data-[hover]:text-tabs-tab-text data-[hover]:border-b-tabs-tab-hover data-[hover]:border-b-8 data-[hover]:border-solid disabled:text-tabs-tab-disabled selected:border-b-primary-600 selected:border-b-8 selected:border-solid px-2 pb-2 text-lg"`
  );
});

test('supports disabled prop', () => {
  render(
    <Tabs disabledKeys={['2']}>
      <Tabs.Item title="tab1" key={1}>
        tab-1 content
      </Tabs.Item>
      <Tabs.Item title="tab2" key={2}>
        tab-2 content
      </Tabs.Item>
    </Tabs>
  );
  const tab = screen.getByText('tab2');
  expect(tab).toHaveAttribute('aria-disabled');
  fireEvent.click(tab);
  expect(screen.getByText('tab-1 content')).toBeVisible();
});

test('set defaultValue via props in tabs', () => {
  render(
    <Tabs defaultSelectedKey={'2'}>
      <Tabs.Item key={1} title="tab1">
        tab-1 content
      </Tabs.Item>
      <Tabs.Item key={2} title="tab2">
        tab-2 content
      </Tabs.Item>
    </Tabs>
  );
  expect(screen.getByText('tab-2 content')).toBeVisible();
});

test('cursor indicates interactivity', () => {
  render(
    <Tabs disabledKeys={['2']}>
      <Tabs.Item key="1" title={'tab1'}>
        tab-1 content
      </Tabs.Item>
      <Tabs.Item key="2" title={'tab2'}>
        tab-2 content
      </Tabs.Item>
    </Tabs>
  );
  const tabs = screen.getAllByRole('tab');
  expect(tabs[0].className).toMatchInlineSnapshot(
    `"flex cursor-pointer justify-center aria-disabled:cursor-not-allowed min-h-[40px] data-[hover]:text-tabs-tab-text data-[hover]:border-b-tabs-tab-hover data-[hover]:border-b-8 data-[hover]:border-solid disabled:text-tabs-tab-disabled selected:border-b-primary-600 selected:border-b-8 selected:border-solid px-2 pb-2 text-lg"`
  );
  expect(tabs[1].className).toMatchInlineSnapshot(
    `"flex cursor-pointer justify-center aria-disabled:cursor-not-allowed min-h-[40px] data-[hover]:text-tabs-tab-text data-[hover]:border-b-tabs-tab-hover data-[hover]:border-b-8 data-[hover]:border-solid disabled:text-tabs-tab-disabled selected:border-b-primary-600 selected:border-b-8 selected:border-solid px-2 pb-2 text-lg"`
  );
});

test('open tabpanel when its tab controller is clicked', () => {
  render(
    <Tabs>
      <Tabs.Item title="tab1" key="1">
        tab-1 content
      </Tabs.Item>
      <Tabs.Item title="tab2" key="2">
        tab-2 content
      </Tabs.Item>
    </Tabs>
  );
  const tab = screen.getByText('tab2');
  fireEvent.click(tab);
  expect(tab.className).toMatchInlineSnapshot(
    `"flex cursor-pointer justify-center aria-disabled:cursor-not-allowed min-h-[40px] data-[hover]:text-tabs-tab-text data-[hover]:border-b-tabs-tab-hover data-[hover]:border-b-8 data-[hover]:border-solid disabled:text-tabs-tab-disabled selected:border-b-primary-600 selected:border-b-8 selected:border-solid px-2 pb-2 text-lg"`
  );
  expect(screen.getByText('tab-2 content')).toBeVisible();
});
