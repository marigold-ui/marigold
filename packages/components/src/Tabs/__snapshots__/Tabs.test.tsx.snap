// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allow styling TabPanel & container via theme: flex gap-2 1`] = `"flex gap-2 mb-[10px]"`;
